typed-rng/src/lib.rs

```diff
@@ -0,0 +1,234 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Typed RNGs with support for trees seeds.
+//!
+//! ## [`TypedRng`]
+//!
+//! This library contains [`TypedRng`], a simple wrapper around a random number
+//! generator that generates values of a particular type.
+//!
+//! At the moment, it only supports stateless value creation, where the
+//! `Generatable::generate` method does not have access to anything other than
+//! the RNG state. It may be extended in the future with the capability to pass
+//! in persistent state.
+//!
+//! ### Comparison with property-based testing
+//!
+//! In a sense, this is a very simple version of how random values are
+//! generated with property-based testing, with e.g. `Arbitrary` or proptest's
+//! `Strategy`.
+//!
+//! But with property-based tests, the goal is for small changes in the random
+//! input to result in _small_ changes in the output. High-quality libraries
+//! like proptest also focus on shrinking.
+//!
+//! With `Generatable`, the goal is for small changes in the random input (the
+//! seed used to initialize the RNG) to result in _large_ changes in output.
+//! (However, it is possible to delegate the bulk of the operation to a value
+//! generator from a PBT framework). There is also no need for shrinking.
+//!
+//! Overall, this means that `Generatable` can be used in both testing and
+//! production.
+//!
+//! ## Other functionality
+//!
+//! This crate also provides two additional convenience functions:
+//!
+//! - [`from_seed`], which generates a new RNG from a seed.
+//! - [`from_parent_and_seed`], which generates a new RNG from a parent RNG and
+//!   a seed.
+//!
+//! Both these methods are short, but more ergonomic and less prone to misuse
+//! than using the underlying libraries directly.
+
+use std::{fmt, hash::Hash, marker::PhantomData};
+
+use rand::rngs::StdRng;
+use rand_core::{RngCore, SeedableRng};
+use uuid::Uuid;
+
+/// Returns a new RNG generated only from the given seeds `seed` and `extra`.
+/// `seed` may be passed down from another caller, e.g. a test, and `extra`
+/// should be a fixed value specific to the callsite.
+///
+/// This takes two hashable arguments rather than one, because when one is
+/// passing down a seed, it is all too easy to not include any extra
+/// information. That may result in multiple different RNGs generating the same
+/// random values. So we expect that callers will also provide bytes of their
+/// own, specific to the callsite, and gently guide them towards doing the
+/// right thing.
+pub fn from_seed<R, H, H2>(seed: H, extra: H2) -> R
+where
+    R: SeedableRng,
+    H: Hash,
+    H2: Hash,
+{
+    // XXX: is Hash really the right thing to use here? That's what
```

Comment:
The guarantees around stability seem nice in stable-hash, especially if we want to persist these values. However, the backwards compatibility seems dangerous in that it ignores default values. Probably best to stick with this for now.